Refactor for issue #4

joefearnley opened the pull request:
Hi @peter279k,

I've refactored this library quite a bit to accommodate for the TODO issue #4. I extracted all of the services out into separate classes that implement a Service interface. Then using the factory pattern, I created a ServiceFactory class that does the creating of those service class objects. 

The interface has changed a bit. To create a service objects, you would use the ServiceFactory class like so:

```
$config = [
    'api-key' => 'key-1234567890',
    'other-setting' => '1234567890'
];

$exampleService = (new ServiceFactory)->create('name-of-service');
$mailgunService->setConfig($config);
$response = $mailgunService->sendReq();
``` 

See the tests for examples. 

Something else we could do, if you want to go this route, is make the create method static so it would something like this:

`$exampleService = ServiceFactory::create('name-of-service');`

I also created a tests directory and a test for each class. 

If you want to accept this pull request, let me know and I will update the README to work with the this interface.

Thanks!

joefearnley committed:
Adding SendGrid support

joefearnley committed:
Adding example SendGrid config file

joefearnley committed:
Removing SendGrid unit test

joefearnley committed:
Resetting main class

joefearnley committed:
Extracting services into thier own file

joefearnley committed:
Reverting ServiceBundle class back and creating ServiceFactory

joefearnley committed:
Adding tests directory and initial files

joefearnley committed:
Adding Bitly service

joefearnley committed:
Attempting initial tests;

joefearnley committed:
Getting tests for mailgun service running

joefearnley committed:
creating mailgun tests

joefearnley committed:
Adding tests for Imgur service

joefearnley committed:
Moving services into its own directoty

joefearnley committed:
setting up services directory structure

joefearnley committed:
Adding ImageShack Tests

joefearnley committed:
Adding Bit.ly tests

joefearnley committed:
Adding McAfee service tests

joefearnley committed:
Adding Google service and tests

joefearnley committed:
Adding test for unknown service

joefearnley committed:
Formatting a couple of small things

joefearnley committed:
Adding SendGrid tests

peter279k commented on src/serviceBundle/ServiceFactory.php:
```diff
@@ -0,0 +1,43 @@
+{
+    public function create($service) 
+    {
+        switch ($service) {
```
What do you think using the following sample code to replace the switch....case?

```php
$service = 'MailGun';
$object = new $service();
```

joefearnley replied:
Yes. Let me look at the options for replacing the switch statement.

peter279k commented on src/serviceBundle/Services/Bitly.php:
```diff
@@ -0,0 +1,22 @@
+        $client = new \GuzzleHttp\Client();
+        return $client->get($url)->json();
+    }
+}
```
Missing the EOF.

peter279k commented:
Hi @joefearnley, thank you for your PR.
There're some request changes that you need to consider.
Don't be worried about the Travis-CI error, but the coding style(StyleCI) you have to fix this.
And the Travis-CI error I will solve.

Thanks.

joefearnley commented:
OK. I will get these changes made.

Another thing I forgot to mentions in the first comment - I have not figured out how to hide API keys in tests. For the Mailgun and Sendgrid tests I just mocked the API so the tests will pass. Might be a good idea to utilize environment variables for this? Let me know what you think.

joefearnley committed:
Removing duplicate class and adding OEF lines for all files

peter279k commented:
Hi @joefearnley, thank you for your reply.
The API testing are two approaches.
Using the Mock to pass the HTTP testing and do the real testing.

The mock approach is simple and good, but it cannot guarantee that it will get the real result.
To keep the test simple, I prefer the mock approach from now on.
The services are not little and maybe the bundle will add other service in the future.

joefearnley committed:
Refactoring ServiceFactory class

joefearnley committed:
Changing the Service from an interface to an abstract class

joefearnley commented:
Alright, @peter279k I refactored the ServiceFactory class. It turns out class_exists() needs the namespace. That or I am not doing something right with namespaces. 

Also, I changed the Service interface to an abstract class. That way I could remove the setConfig() method from every service class. 

Oh, and I added a newline to every file. 

Hope this helps!

joefearnley committed:
changing temporary test

peter279k commented:
Hi @joefearnley, thank you for your latest commit.
It looks good for me.
One more thing the StyleCI is still failed during checking coding style.
Would you fix them?
Once fixing coding style, I will merge the PR.

Thanks.

joefearnley committed:
Making changes for style ci

joefearnley commented:
Hi @peter279k. I agree, mocking the API calls certainly do not guarantee that the results are correct. I'll mock the calls from now on until you specify otherwise.

I _think_ I have made all of the changes style ci requested. If not, please let me know.

peter279k commented:
Great! @joefearnley, thank you for this PR.
If you want to improve the testing, please feel free to create another PR.

Thanks.
